### PR TITLE
Support for jobs triggered on failure of other jobs (#140)

### LINKED ISSUE #140: Adding on success and on failure in dependent jobs
A cool feature in my opinion would be having an optional on failure jobs when creating a job, so you can have a create flow of what happens if the current job failed (maybe clean up jobs, etc..)

### PR DESCRIPTION
**Pull request by itstehkman**
Fixes #140. This PR allows you to define a job that gets triggered on failure of another job by setting OnFailureJob for a given Job. The test cases I've provided make sure that 1) this on failure job gets triggered on failure of the parent job 2) it does not get triggered when the parent job succeeds.

Some questions that may need to be answered and resolved are:
1) What happens if the on failure job gets triggered while it is already happening? Maybe we should specify a specific job type for on failure jobs only that disallows this? Or maybe we could just wait for the job to finish via its lock, and then run it when its done with its current run
2) How do you create these on failure jobs? Currently, I've shown via the test cases that you can add this job via the API, which may cause it to run on init depending on the schedule parameter. I do not feel like this is desired functionality, but I'll leave it up to you decide. I believe the desired functionality is that you create a job that never runs unless triggered on failure.

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -137,6 +137,7 @@ func HandleAddJob(cache job.JobCache, defaultOwner string) func(http.ResponseWri
 
 		err = newJob.Init(cache)
 		if err != nil {
+			// TODO(itstehkman): should we run on_failure job here?
```
Comment (ajvb) [inline api/api.go]: I'd say not. If you can't initialize, that usually means something fundamentally wrong.

File: job/job.go [hunk 2]
```diff
@@ -409,14 +412,31 @@ func (j *Job) DeleteFromDependentJobs(cache JobCache) error {
...
+}
+
 func (j *Job) Run(cache JobCache) {
 	// Schedule next run
 	j.lock.RLock()
 	jobRunner := &JobRunner{job: j, meta: j.Metadata}
 	j.lock.RUnlock()
 	newStat, newMeta, err := jobRunner.Run(cache)
 	if err != nil {
+		// TODO: run on_failure job
```
Comment (ajvb) [inline job/job.go]: is this TODO needed?

File: job/job.go [hunk 3]
```diff
@@ -409,14 +412,31 @@ func (j *Job) DeleteFromDependentJobs(cache JobCache) error {
 	return nil
 }
 
+// Runs the on failure job, if it exists. Does not lock the job - it is up to you to do this
```
Comment (ajvb) [inline job/job.go]: We should lock the reading of the OnFailureJob

File: job/job.go [hunk 4]
```diff
@@ -493,3 +513,10 @@ func (j *Job) MarshalJSON() ([]byte, error) {
 	defer j.lock.RUnlock()
 	return json.Marshal(RJob(*j))
 }
+
+// TODO(itstehkman): This could go in j.Metadata
```
Comment (ajvb) [inline job/job.go]: That's a good idea, I'd suggest we move it there.

File: job/job_test.go [hunk 5]
```diff
@@ -975,3 +975,62 @@ func TestRemoteJobBadStatusSuccess(t *testing.T) {
...
+	for {
+		j.lock.RLock()
+		if j.IsDone {
+			j.lock.RUnlock()
+			break
+		}
+		j.lock.RUnlock()
+		time.Sleep(1)
+	}
+
```
Comment (ajvb) [inline job/job_test.go]: extra space?

File: job/job_test.go [hunk 6]
```diff
@@ -975,3 +975,62 @@ func TestRemoteJobBadStatusSuccess(t *testing.T) {
...
+	j := GetMockJob()
+	j.OnFailureJob = onFailureJob.Id
+
+	j.Init(cache)
+
+	waitForJob(j)
+
+	j.lock.RLock()
+	onFailureJob.lock.RLock()
+	assert.Equal(t, j.Metadata.SuccessCount, uint(1))
```
Comment (ajvb) [inline job/job_test.go]: Want to check for `onFailureJob.Metadata.SuccessCount` as well?

File: job/job.go (hunk 3 above)
Comment (itstehkman) [reply job/job.go]: I believe that is already taken care of in Job.Run, or did you mean something else?

**Comment by itstehkman** [comment]
@ajvb my pleasure to help. This is an awesome project!

**Comment by ajvb** [comment]
@itstehkman Awesome! thanks a bunch

**Pull request merged**